File: boat-sample/src/test/java/sample/java/xyz/srclab/common/run/RunSample.java
package sample.java.xyz.srclab.common.run;

import org.testng.annotations.Test;
import xyz.srclab.common.lang.Current;
import xyz.srclab.common.lang.IntRef;
import xyz.srclab.common.run.Runner;
import xyz.srclab.common.run.Running;
import xyz.srclab.common.run.Scheduler;
import xyz.srclab.common.run.Scheduling;
import xyz.srclab.common.test.TestLogger;

import java.time.Duration;

public class RunSample {

    private static final TestLogger logger = TestLogger.DEFAULT;

    @Test
    public void testRunner() {
        Runner runner = Runner.SYNC_RUNNER;
        IntRef intRef = IntRef.of(0);
        Running<?> running = runner.run(() -> {
            intRef.set(666);
            return null;
        });
        running.get();
        //666
        logger.log("int: {}", intRef.get());
    }

    @Test
    public void testScheduledRunner() {
        Scheduler scheduler = Scheduler.DEFAULT_THREAD_SCHEDULER;
        IntRef intRef = IntRef.of(0);
        Scheduling<?> scheduling = scheduler.scheduleFixedDelay(Duration.ZERO, Duration.ofMillis(1000), () -> {
            intRef.set(intRef.get() + 100);
            return null;
        });
        Current.sleep(2500);
        scheduling.cancel(false);
        //300
        logger.log("int: {}", intRef.get());
    }
}

File: boat-sample/src/test/kotlin/sample/xyz/srclab/common/run/RunSample.kt
package sample.xyz.srclab.common.run

import org.testng.annotations.Test
import xyz.srclab.common.base.Current.sleep
import xyz.srclab.common.base.IntRef.Companion.of
import xyz.srclab.common.run.Runner
import xyz.srclab.common.run.Running
import xyz.srclab.common.run.ScheduledRunner
import xyz.srclab.common.run.ScheduledRunning
import xyz.srclab.common.test.TestLogger
import java.time.Duration

class RunSampleKt {

    @Test
    fun testRunner() {
        val runner: Runner = Runner.SYNC_RUNNER
        val intRef = of(0)
        val running: Running<*> = runner.run<Any?> {
            intRef.set(666)
            null
        }
        running.get()
        //666
        logger.log("int: {}", intRef.get())
    }

    @Test
    fun testScheduledRunner() {
        val runner = ScheduledRunner.SINGLE_THREAD_RUNNER
        val intRef = of(0)
        val running: ScheduledRunning<*> = runner.scheduleWithFixedDelay<Any?>(Duration.ZERO, Duration.ofMillis(1000)) {
            intRef.set(intRef.get() + 100)
            null
        }
        sleep(2500)
        running.cancel(false)
        //300
        logger.log("int: {}", intRef.get())
    }

    companion object {
        private val logger = TestLogger.DEFAULT
    }
}